source/model-generator/AnnotationFeatureTemplate.cpp

```diff
@@ -0,0 +1,130 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <DexAnnotation.h>
+#include <fmt/format.h>
+#include <mariana-trench/Context.h>
+#include <mariana-trench/FeatureFactory.h>
+#include <mariana-trench/JsonValidation.h>
+#include <mariana-trench/Method.h>
+#include <mariana-trench/model-generator/AnnotationFeatureTemplate.h>
+
+namespace marianatrench {
+
+AnnotationTarget::AnnotationTarget(Kind kind,
+                                   std::optional<RootTemplate> argument)
+    : kind_(kind), argument_(std::move(argument)) {}
+
+AnnotationTarget AnnotationTarget::from_json(const Json::Value& value) {
+  const std::string target = JsonValidation::string(value);
+  Kind kind;
+  std::optional<RootTemplate> argument;
+  if ("Class" == target) {
+    kind = Kind::Class;
+  } else if ("Method" == target) {
+    kind = Kind::Method;
+  } else {
+    kind = Kind::Parameter;
+    try {
+      argument = RootTemplate::from_json(value);
+    } catch (...) {
+      throw_with_nested(JsonValidationError(
+          value,
+          /* field */ std::nullopt,
+          fmt::format("valid annotation target (`Class`, `Method`, or "
+                      "`Argument(...)`), got `{}`",
+                      target)));
+    }
+  }
+  return AnnotationTarget{kind, std::move(argument)};
+}
+
+AnnotationTarget::Kind AnnotationTarget::kind() const { return kind_; }
+
+const RootTemplate& AnnotationTarget::argument() const { return *argument_; }
+
+AnnotationFeatureTemplate::AnnotationFeatureTemplate(
+    AnnotationTarget target, const DexType* dex_type,
+    const DexString* MT_NULLABLE tag,
+    const DexString* MT_NULLABLE annotation_parameter_name)
+    : target_(std::move(target)),
+      dex_type_(dex_type),
+      tag_(tag),
+      annotation_parameter_name_(annotation_parameter_name) {}
+
+AnnotationFeatureTemplate AnnotationFeatureTemplate::from_json(
+    const Json::Value& value) {
+  JsonValidation::check_unexpected_members(
+      value, {"target", "type", "tag", "annotation_parameter"});
+  AnnotationTarget port = AnnotationTarget::from_json(value["target"]);
+  const DexType* dex_type = JsonValidation::dex_type(value, "type");
+  Json::Value optional_value = value["tag"];
+  const DexString* tag = nullptr;
+  if (!optional_value.isNull()) {
+    tag = DexString::make_string(JsonValidation::string(optional_value));
+  }
+  optional_value = value["annotation_parameter"];
+  const DexString* annotation_parameter_name = nullptr;
+  if (!optional_value.isNull()) {
+    annotation_parameter_name = DexString::make_string(JsonValidation::string(optional_value));
+  }
+
+  return AnnotationFeatureTemplate{std::move(port), dex_type, tag, annotation_parameter_name};
+}
+
+const Feature* MT_NULLABLE AnnotationFeatureTemplate::instantiate(
+    const Method* method, Context& context,
+    const TemplateVariableMapping& parameter_positions) const {
+  const DexAnnotationSet* anno_set;
```

Comment:
nit: Mixing `return` in the switch is a bit confusing to read. I'd say:
- initialize to `nullptr` here: 
`  const DexAnnotationSet* anno_set = nullptr;`
- update value in switch
- Add early return case after switch:
```
if (anno_set == nullptr) {
  return nullptr;
}
```